File: DMOPC '17 Contest 1 P4 - Quests.cpp
#include <bits/stdc++.h>
#define int long long

using namespace std;

const int inf = 1LL << 60;

signed main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, H;
    cin >> N >> H;
    vector<int> prev1(H + 1, -inf);  // did not take previous set
    vector<int> prev2(H + 1, -inf);

    prev1[0] = prev2[0] = 0;

    // (val, cost), (val, cost)
    for (int i = 0; i < N; i++) {
        int v1, c1, v2, c2;
        cin >> v1 >> c1 >> v2 >> c2;

        // 0/1 first for going to the NPCs
        vector<int> dp2(H + 1, -inf);
        vector<int> dp1(H + 1);
        for (int j = 0; j <= H; j++) {
            dp1[j] = max(prev1[j], prev2[j]);  // not taking from this set
        }

        for (int v = H; v >= c1; v--) {
            dp2[v] = dp1[v - c1] + v1;
        }

        // unbounded for doing amounts of quests
        for (int v = c2; v <= H; v++) {
            dp2[v] = max(dp2[v], dp2[v - c2] + v2);
        }

        prev1 = dp1;
        prev2 = dp2;
    }

    cout << max(*max_element(prev1.begin(), prev1.end()), *max_element(prev2.begin(), prev2.end())) << "\n";

    return 0;
}

File: DMOPC '17 Contest 1 P4 - Quests.py
# TLE, CHECK C++ CODE
# https://dmoj.ca/problem/dmopc17c1p4
# Mixed knapsack
# Part 1: 0/1 knapsack
# Part 2: unbounded knapsack
#
# TC: O(N * H)

def solve():
    inf = 1 << 60
    N, H = map(int, input().split())
    prev1 = [-inf] * (H + 1)  # did not take previous set
    prev2 = [-inf] * (H + 1)

    prev1[0] = prev2[0] = 0

    # (val, cost), (val, cost)
    for _ in range(N):
        v1, c1, v2, c2 = map(int, input().split())

        # 0/1 first for going to the NPCs
        dp2 = [-inf] * (H + 1)
        dp1 = [max(i, j) for i, j in zip(prev1, prev2)]  # not taking from this set

        for v in reversed(range(c1, H + 1)):
            dp2[v] = dp1[v - c1] + v1

        # unbounded for doing amounts of quests
        for v in range(c2, H + 1):
            dp2[v] = max(dp2[v], dp2[v - c2] + v2)

        prev1 = dp1
        prev2 = dp2

    print(max(max(prev1), max(prev2)))


solve()

"""
2 7
9 1 7 1
5 3 2 1
7 1 4 3
1 2 8 1
"""
